### DISCUSSION TITLE
Does the runtime folder contain actual runtimes, or loaders for the runtimes?

### DISCUSSION THREAD

**Discussion by protegesolutions** [MicrosoftEdge/WebView2Feedback#3972] [Q&A]
I'm kinda confused by this. The compiler emits a folder called _runtimes_ which contains sub-folders for three machine architectures (win-x86, win-x64 and win-arm64) which each contain a _native_ sub-folder within which is are files _WebView2Loader.dll_.

Now, the doco says that the WebView2 runtimes must be installed on the target machines and that these are usually installed in _C:\Program Files (x86)\Microsoft\EdgeWebView\Application_ and the install state can be checked using registry checks (or an API).

So ... which is true? That the _WebView2Loader.dll_ is an _actual_ runtime and doesn't need to be deployed/distributed if the runtimes are already installed in the _EdgeWebView_ folder? Or that these are not runtimes and are actually loaders _for_ the runtimes?

**Comment by Syul968** [comment]
Hi @protegesolutions,

The `WebView2Loader.dll` files are, like their name suggest, loaders for the _WebView2 Runtime_. They happen to be placed under a directory called "runtimes", but that's just details about the build output. The loader needs to be distributed with your application since it will find the WebView2 Runtime install for your app to use WebView2.

Hope this helps,
Luis

**Comment by protegesolutions** [comment]
Thanks for your reply. In our case, we are trying to replace the _WebBrowser_ control, due to increasing incompatibilities, in a class library that is consumed by third-party apps, services, web sites, etc. We prefer to keep the library as a single file dll, so we embed references. We can't embed the _WebView2Loader.dll_ files (for two architectures at least) because they are native code. This over-complicates things enormously for us.

Also, the nomenclature of the various code bundles is a bit confusing; we have WebView2Loader contained in a runtimes folder, the _WebView2.Core/WinForms/Wpf_ files referred to as runtimes and the actual _EdgeWebView_ runtimes.

**Follow up question:** why are the loaders separate from the actual libraries (_WebView2.Core/WinForms/Wpf_)? Could they not be sequestered within the libraries, thereby obviating the need to distribute them?

I realize that the main problem (for us) is the fact that we are targeting a class library, rather than a desktop app or a service or web app, etc. But I have actually seen references from others asking about class library support. The _WebBrowser_ control gave us a very easy and seamless solution whereas _WebView2_ ... well, I've spent several days trying to resolve various issues, so far to no avail.

**Comment by Syul968** [comment]
The WebView2 Loader is a small architecture-specific component that helps apps locate the WebView2 Runtime. Since CLR assemblies can run in a variety of architectures, all WebView2 Loader DLLs are required to ship with your app/component so it works across different devices. The .NET assembly for WebView2 relies on these DLLs shipping with the app/component, in the layout detailed in our documentation: [Files to ship with the app](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution#files-to-ship-with-the-app).

For clarity, these are the different pieces you're referring to:

* `WebView2Loader`. A small component that helps apps locate the WebView2 Runtime.
* `Microsoft.Web.WebView2.Core`. "Core" .NET assembly of the WebView2 API.
* `Microsoft.Web.WebView2.WinForms`/`WPF`. WebView2 controls for the WinForms and WPF frameworks.
* `WebView2 Runtime`. A redistributable runtime that serves as the underlying web platform for WebView2 apps

Have you tried embedding the loader DLLs as resources? Please note that you might also need to extract it during initialization of your component. Also, [CoreWebView2Environment.SetLoaderDllFolderPath](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2environment.setloaderdllfolderpath?view=webview2-dotnet-1.0.1293.44) lets you set the directory to look for the WebView2Loader DLL matching the architecture of the device.
